types/src/account_address.rs

```diff
@@ -50,8 +47,12 @@ impl AccountAddress {
     }
 
     pub fn from_public_key<PublicKey: VerifyingKey>(public_key: &PublicKey) -> Self {
-        let hash = *HashValue::from_sha3_256(&public_key.to_bytes()).as_ref();
-        AccountAddress::new(hash)
+        let hash = HashValue::from_sha3_256(&public_key.to_bytes()).to_vec();
+        // keep only last 16 bytes
+        let mut array = [0; ADDRESS_LENGTH];
+        // TODO: make sure this is what we want
+        array.copy_from_slice(&hash[ADDRESS_LENGTH..32]);
```

Comment:
If we decide not to double hash the PK, I think it’s safer to keep the first 16 bytes, so it’s independent to the output size of hash. 

That said if we ever update our hash function to SHA384 or SHA512, then [0..ADDRESS_LENGTH] will still work, while [ADDRESS_LENGTH..32] is valid, but it’s not the last bytes of the hash.

Also, is double hashing out of scope due to performance reasons? If we do, then you cannot easily apply land grabbing to short addresses, but it’s true you can still do for regular ones.

Reply:
Good point; first 16 bytes makes more sense from a protocol evolution/crypto agility perspective. I wonder why Eth decided to keep the rightmost 16?

Re: double hashing, I definitely intend to do it, but will write the equivalent Move code before updating this.

Reply:
> I wonder why Eth decided to keep the rightmost 16?
I have no idea and security-wise it's the same; actually they kept 20, we are more "succinct" :)

I was also thinking if we should add some SALT, i.e., `truncate_first16(hash(libra_salt, pk))` , mainly to reduce the vector of potential future collision attacks by using generic precomputed rainbow tables. It's true, that unlike other blockchains, Libra's addresses are decoupled from PKs and they are just identifiers. But adding an extra Libra-unique way of generating hashes does not harm.

Reply:
So: 
- I did not do double hashing here because we didn't do it before, and changing it at the same time as changing the address format turned out to be very difficult. So the current scheme still allows land grabs. We can fix this in a follow-up.
- It turns out that keeping the leftmost 16 byte is difficult because the Move source language and IR let you write (e.g.) `0xa550c` and pad the result with *leading* zeroes. Changing this will break a ton of things, so I decided to keep the rightmost 16 bytes for now and we can reconsider later (in parallel with changes to the source lang.
- Adding a salt may also be a good idea, but I think we should consider it later to minimize the breakages/churn of this diff.